sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/internal/view/FilteredAttributes.java

```diff
@@ -0,0 +1,265 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.sdk.metrics.internal.view;
+
+import io.opentelemetry.api.common.AttributeKey;
+import io.opentelemetry.api.common.Attributes;
+import io.opentelemetry.api.common.AttributesBuilder;
+import io.opentelemetry.api.internal.ImmutableKeyValuePairs;
+import java.util.BitSet;
+import java.util.Collections;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.StringJoiner;
+import java.util.function.BiConsumer;
+import javax.annotation.Nullable;
+
+/**
+ * Filtered attributes is a filtered view of a {@link ImmutableKeyValuePairs} backed {@link
+ * Attributes} instance. Rather than creating an entirely new attributes instance, it keeps track of
+ * which source attributes are excluded while implementing the {@link Attributes} interface.
+ *
+ * <p>Notably, the {@link FilteredAttributes#equals(Object)} and {@link
+ * FilteredAttributes#hashCode()} depend on comparison against other {@link FilteredAttributes}
+ * instances. This means that where {@link FilteredAttributes} is used for things like map keys, it
+ * must be used for all keys in that map. You cannot mix {@link Attributes} implementations. This is
+ * also true for the default attributes implementation.
+ */
+@SuppressWarnings("unchecked")
+final class FilteredAttributes implements Attributes {
+
+  private static final int BITS_PER_INTEGER = 32;
+
+  // Backing source data from ImmutableKeyValuePairs.data. This array MUST NOT be mutated.
+  private final Object[] sourceData;
+  // The bits of filteredIndices track whether the first 32 key-values of sourceData should be
+  // excluded in the output. A bit equal to 1 indicates the sourceData key at i*2 should be
+  // excluded. overflowFilteredIndices is used when more than 32 key-value pairs are present in
+  // sourceData.
+  private final int filteredIndices;
+  @Nullable private final BitSet overflowFilteredIndices;
```

Comment:
I experimented with all sorts of different mechanisms to track which attributes from the original source data should be filtered out and landed this. We use a `int` to track the first 32 key value pairs, which covers most attribute use cases we'll see in the wild. For larger attribute sets, we overflow into `BitSet`, which is less efficient for typical use cases.